File: ts/prereqs/transfer.ts
import {
  Transaction,
  SystemProgram,
  Connection,
  Keypair,
  LAMPORTS_PER_SOL,
  sendAndConfirmTransaction,
  PublicKey,
} from "@solana/web3.js";
import wallet from "../wallet";
import dotenv from "dotenv";

// Load environment variables
dotenv.config();

// Reconstruct keypair from wallet file
const from = Keypair.fromSecretKey(new Uint8Array(wallet));

// Define the recipient
const to = new PublicKey(
  process.env.WALLET_1_ADDRESS || "YOUR_WALLET_ADDRESS_HERE"
);

// Setup connection
const connection = new Connection("https://api.devnet.solana.com");

(async () => {
  try {
    // Create transaction to send 0.1 SOL
    const transaction = new Transaction().add(
      SystemProgram.transfer({
        fromPubkey: from.publicKey,
        toPubkey: to,
        lamports: 2 * LAMPORTS_PER_SOL,
      })
    );

    // Set recent blockhash & fee payer
    transaction.recentBlockhash = (
      await connection.getLatestBlockhash("confirmed")
    ).blockhash;
    transaction.feePayer = from.publicKey;

    // Sign, send, and confirm
    const signature = await sendAndConfirmTransaction(connection, transaction, [
      from,
    ]);

    console.log(
      `✅ Success! View your TX: https://explorer.solana.com/tx/${signature}?cluster=devnet`
    );
  } catch (e) {
    console.error(`❌ Error: ${e}`);
  }
})();

File: 1_rs/src/transfer.rs
use solana_client::rpc_client::RpcClient;
use solana_program::{pubkey::Pubkey, system_instruction::transfer};
use solana_sdk::{
    signature::{Signer, read_keypair_file},
    transaction::Transaction,
    hash::hash,
};
use std::str::FromStr;
use std::env;
use dotenvy::dotenv;

const RPC_URL: &str = "https://api.devnet.solana.com";

pub fn main() {
    // Load your devnet keypair from file
    dotenv().ok();
    let wallet_path = env::var("WALLET_3").expect("WALLET not set");
    let keypair = read_keypair_file(wallet_path).expect("Couldn't read keypair");
    let pubkey = keypair.pubkey();

    // Optional: sign and verify a message
    let message_bytes = b"I verify my Solana Keypair!";
    let sig = keypair.sign_message(message_bytes);
    let sig_hashed = hash(sig.as_ref());
    match sig.verify(pubkey.as_ref(), sig_hashed.as_ref()) {
        true => println!("Signature verified"),
        false => println!("Verification failed"),
    }

    // Define recipient pubkey (replace with your target public key)
    let to_pubkey = Pubkey::from_str(&env::var("WALLET_2_ADDRESS").unwrap())
        .expect("Invalid WALLET_2_ADDRESS");

    // Create RPC client
    let rpc_client = RpcClient::new(RPC_URL.to_string());

    // Get recent blockhash for transaction
    let recent_blockhash = rpc_client.get_latest_blockhash().expect("Failed to get recent blockhash");

    // Create transfer instruction (sending 1000 lamports = 0.000001 SOL)
    let transfer_instruction = transfer(&pubkey, &to_pubkey, 500_000_000);

    // Build and sign transaction
    let transaction = Transaction::new_signed_with_payer(
        &[transfer_instruction],
        Some(&pubkey),
        &[&keypair],
        recent_blockhash,
    );

    // Send and confirm transaction
    let signature = rpc_client
        .send_and_confirm_transaction(&transaction)
        .expect("Failed to send transaction");

    println!("Success! Check your TX here:");
    println!("https://explorer.solana.com/tx/{}?cluster=devnet", signature);
}